File: 0095-unique-binary-search-trees-ii/0095-unique-binary-search-trees-ii.cpp

class Solution {
public:
    map<pair<int, int>, vector<TreeNode*>> mp;
    vector<TreeNode*> solve(int start, int end){
        
        if(start > end){
            return { NULL };
        }
        
        if(mp.find({start, end}) != mp.end())
            return mp[{start, end}];
        
        vector<TreeNode*> result;
        
        for(int i=start; i<=end; i++){
            
            vector<TreeNode*> leftBSTs = solve(start, i-1);
            vector<TreeNode*> rightBSTs = solve(i+1, end);
            
            for(auto leftRoot: leftBSTs){
                for(auto rightRoot: rightBSTs){
                    
                    TreeNode* root = new TreeNode(i);
                    
                    root->left = leftRoot;
                    root->right = rightRoot;
                    
                    result.push_back(root);
                }
            }
        }
        
        return mp[{start, end}] = result;
    }
    
    vector<TreeNode*> generateTrees(int n) {
        
        return solve(1, n);
    }
};

File: 0095-unique-binary-search-trees-ii/0095-unique-binary-search-trees-ii.java
class Solution {
    public List<TreeNode> solve(int start, int end){
        
        if(start > end){
            List<TreeNode> l = new ArrayList<>();
            l.add(null);
            return l;
        }
        
        List<TreeNode> result = new ArrayList<>();
        
        for(int i=start; i<=end; i++){
            
            List<TreeNode> leftBSTs = solve(start, i-1);
            List<TreeNode> rightBSTs = solve(i+1, end);
            
            for(TreeNode leftRoot: leftBSTs){
                for(TreeNode rightRoot: rightBSTs){
                    
                    TreeNode root = new TreeNode(i);
                    root.left = leftRoot;
                    root.right = rightRoot;
                    result.add(root);
                }
            }
        }
        return result;
    }
    
    public List<TreeNode> generateTrees(int n) {
        
        return solve(1, n);
    }
}